---
title: "xCell2 custom bulk reference tutorial"
author: "Almog Angel"
date: '2023-04-10'
output: html_document
---

```{r}
library(tidyverse)
```

-   **Load "BlueprintEncodeData" reference via `celldex` package**

```{r}
bp <- celldex::BlueprintEncodeData() 
```

-   **Convert the gene expression table to a matrix**

```{r}
bp_ref <- as.matrix(bp@assays@data$logcounts)
bp_ref[1:5, 1:5]
```

-   **Load cell type labels**

```{r}
bp_labels <- bp@colData
head(bp_labels)
```

-   **Convert labels to xCell2.0 format:**

1.  The first column "ont" for cell type [ontology](https://www.ebi.ac.uk/ols/index) (i.e., "CL:0000775", "CL:0000576") **\*\***

2.  The second column "label" can be any string that describes the cell type (i.e., "Neutrophils", "Monocytes") \*\*\*

3.  The third column "sample" is the sample name in the reference - it must be the same as the reference column names `all(bp_labels$sample == colnames(bp_ref))`

4.  The fourth column "dataset" is the name of the reference/dataset the sample originate from

\*\* Use NA if you don't know the cell type ontology

\*\*\* You can use any label you like but it is recommended to use the same labels as in [ontology](https://www.ebi.ac.uk/ols/index)

```{r}
bp_labels <- bp_labels %>%
  as_tibble() %>% 
  rename("label.ont" = "ont",
         "label.fine" = "label") %>%
  mutate(sample = rownames(bp_labels),
         dataset = "Blueprint-Encode") %>%
  select(ont, label, sample, dataset)
```

-   **Convert labels to a dataframe class and check the samples match column names in reference**

```{r}
bp_labels <- as.data.frame(bp_labels)
all(colnames(bp_ref) == bp_labels$sample)

```

-   **Use `xCell2GetLineage` to check for cell types dependencies**

```{r}

source("https://raw.githubusercontent.com/N-Cohen/xCell2/master/R/xCell2GetLineage.R")
library(ontoProc)
xCell2GetLineage(bp_labels, out_file = "bp_dependencies.tsv")
```

-   **Read the dependencies file and check and edit it manually if you like**

```{r}
setwd("/Users/noam/Desktop/Technion/bioinformatics project/xCell2")
dep_table <- read.table("bp_dependencies.tsv", sep = "\t", header = TRUE)
View(dep_table)
```

-   **Now we are ready to generate custom signatures!**

```{r}
source("https://raw.githubusercontent.com/N-Cohen/xCell2/master/R/xCell2Train.R")
xcell2sigs_bp <- xCell2Train(ref = bp_ref, labels = bp_labels, lineage_file_checked = "bp_dependencies.tsv", data_type = "rnaseq")
```
